flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java

```diff
@@ -272,11 +287,161 @@ private static ShuffleDescriptor createShuffleDescriptor(
+		private final ConnectionManager connectionManager;
+		private final ResultPartitionManager partitionManager;
+		private final TaskEventPublisher taskEventPublisher;
+		private final NetworkBufferPool networkBufferPool;
+
+		public TestSingleInputGateFactory(
+				@Nonnull ResourceID taskExecutorResourceId,
+				@Nonnull NettyShuffleEnvironmentConfiguration networkConfig,
+				@Nonnull ConnectionManager connectionManager,
+				@Nonnull ResultPartitionManager partitionManager,
+				@Nonnull TaskEventPublisher taskEventPublisher,
+				@Nonnull NetworkBufferPool networkBufferPool) {
+			super(
+				taskExecutorResourceId,
+				networkConfig,
+				connectionManager,
+				partitionManager,
+				taskEventPublisher,
+				networkBufferPool);
+			this.networkBufferPool = networkBufferPool;
+			this.taskEventPublisher = taskEventPublisher;
+			this.partitionManager = partitionManager;
+			this.connectionManager = connectionManager;
+			this.partitionRequestMaxBackoff = networkConfig.partitionRequestMaxBackoff();
+			this.partitionRequestInitialBackoff = networkConfig.partitionRequestInitialBackoff();
+			this.taskExecutorResourceId = taskExecutorResourceId;
+		}
+
+		@Override
+		protected InputChannel createKnownInputChannel(
+			SingleInputGate inputGate,
+			int index,
+			NettyShuffleDescriptor inputChannelDescriptor,
+			ChannelStatistics channelStatistics,
+			InputChannelMetrics metrics) {
+			ResultPartitionID partitionId = inputChannelDescriptor.getResultPartitionID();
+			if (inputChannelDescriptor.isLocalTo(taskExecutorResourceId)) {
+				return new TestLocalInputChannel(
+					inputGate,
+					index,
+					partitionId,
+					partitionManager,
+					taskEventPublisher,
+					partitionRequestInitialBackoff,
+					partitionRequestMaxBackoff,
+					metrics);
+			} else {
+				return new TestRemoteInputChannel(
+					inputGate,
+					index,
+					partitionId,
+					inputChannelDescriptor.getConnectionId(),
+					connectionManager,
+					partitionRequestInitialBackoff,
+					partitionRequestMaxBackoff,
+					metrics,
+					networkBufferPool);
+			}
+		}
+	}
+
+	/**
+	 * A {@link LocalInputChannel} which ignores the given subpartition index and uses channel index
+	 * instead when requesting subpartition.
+	 */
+	private static class TestLocalInputChannel extends LocalInputChannel {
+
+		private ResultPartitionID newPartitionID = new ResultPartitionID();
+
+		public TestLocalInputChannel(
+				SingleInputGate inputGate,
+				int channelIndex,
+				ResultPartitionID partitionId,
+				ResultPartitionManager partitionManager,
+				TaskEventPublisher taskEventPublisher,
+				int initialBackoff,
+				int maxBackoff,
+				InputChannelMetrics metrics) {
+			super(
+				inputGate,
+				channelIndex,
+				partitionId,
+				partitionManager,
+				taskEventPublisher,
+				initialBackoff,
+				maxBackoff,
+				metrics);
+		}
+
+		@Override
+		public void requestSubpartition(int subpartitionIndex) throws IOException, InterruptedException {
```

Comment:
In addition, we might also need to override the method of `retriggerSubpartitionRequest`, in case this process might be used to cause invalid `subpartitionIndex` future.

Reply:
Can we always call ```createResultPartitionWriter``` before call ```createReceiver```, which means the result partitions are always registered before requested. Current implementation can guarantee that. I am saying this because I need to override more than one method and also need do some refactor to support ```retriggerSubpartitionRequest```.

Reply:
I add a comment to the ```createReceiver``` method.